File: Sherlock-and-Anagrams.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'sherlockAndAnagrams' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING s as parameter.
 */

int sherlockAndAnagrams(string s)
{
    int result = 0;
    std::unordered_map<string, int> frequencyMap;

    for (int i = 0; i < s.size(); ++i)
    {
        for (int j = i; j < s.size(); ++j)
        {
            string sub = s.substr(i, j - i + 1);
            sort(sub.begin(), sub.end());
            frequencyMap[sub]++;
        }
    }

    for (const auto &pair : frequencyMap)
    {
        int count = pair.second;
        result += (count * (count - 1)) / 2;
    }

    return result;
}


int sherlockAndAnagrams2(string s)
{
    int result = 0;
    std::map<vector<int>, int> frequencyMap;

    for (int i = 0; i < s.size(); ++i)
    {
        vector<int> arr(26, 0);
        
        for (int j = i; j < s.size(); ++j)
        {
            arr[s[j] - 'a']++;
            frequencyMap[arr]++;
        }
    }

    for (const auto& [charFrequency, count] : frequencyMap)
    {
        result += (count * (count - 1)) / 2;
    }

    return result;
}


int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string q_temp;
    getline(cin, q_temp);

    int q = stoi(ltrim(rtrim(q_temp)));

    for (int q_itr = 0; q_itr < q; q_itr++) {
        string s;
        getline(cin, s);

        int result = sherlockAndAnagrams(s);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

File: Sherlock-and-Anagrams.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {

    /*
     * Complete the 'sherlockAndAnagrams' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts STRING s as parameter.
     */

private static int sherlockAndAnagrams_ByString(String s) {
        int result = 0;
        Map<String, Integer> frequencyMap = new HashMap<>();

        for (int i = 0; i < s.length(); ++i) {
            for (int j = i; j < s.length(); ++j) {
                char[] subChars = s.substring(i, j + 1).toCharArray();
                Arrays.sort(subChars);
                String sub = new String(subChars);
                frequencyMap.put(sub, frequencyMap.getOrDefault(sub, 0) + 1);
            }
        }

        for (int count : frequencyMap.values()) {
            result += (count * (count - 1)) / 2;
        }

        return result;
    }

    private static int sherlockAndAnagrams_ByArray(String s) {
        int result = 0;
        Map<String, Integer> frequencyMap = new HashMap<>();

        for (int i = 0; i < s.length(); ++i) {
            int[] arr = new int[26];

            for (int j = i; j < s.length(); ++j) {
                arr[s.charAt(j) - 'a']++;
                String charFrequency = Arrays.toString(arr);
                frequencyMap.put(charFrequency, frequencyMap.getOrDefault(charFrequency, 0) + 1);
            }
        }

        for (int count : frequencyMap.values()) {
            result += (count * (count - 1)) / 2;
        }

        return result;
    }

    public static int sherlockAndAnagrams(String s) {
        if (s.length() < 50) {
            return sherlockAndAnagrams_ByString(s);
        } else {
            return sherlockAndAnagrams_ByArray(s);
        }
    }
}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int q = Integer.parseInt(bufferedReader.readLine().trim());

        IntStream.range(0, q).forEach(qItr -> {
            try {
                String s = bufferedReader.readLine();

                int result = Result.sherlockAndAnagrams(s);

                bufferedWriter.write(String.valueOf(result));
                bufferedWriter.newLine();
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });

        bufferedReader.close();
        bufferedWriter.close();
    }
}
